### ISSUE TITLE
[question] Order of execution of bootstrap providers

### ISSUE THREAD

**Issue by handlespeed** [dotnet/orleans#1867]
We use tag BootstrapProviders in OrleansConfiguration.xml to have several bootStrapProviders. What is the order of initialization of those boot strap providers? I just have a look at Orleans code and got:
`public async Task InitProviders(IProviderRuntime providerRuntime)
        {
            Dictionary<string, TProvider> copy; 
            lock (providers)
            {
                copy = providers.ToDictionary(p => p.Key, p => p.Value);
            }

```
        foreach (var provider in copy)
        {
            string name = provider.Key;
            try
            {
                await provider.Value.Init(provider.Key, providerRuntime, providerConfigs[name]);`
```

so that means, the order is undefined because of truth of type Dictionary?
If that, what if those boot strap providers have dependency order issue each other? Or we have to move all the providers implementation to only one provider.
Thanks.

**Comment by gabikliot** [comment]
Yes, I believe the order is undefined now.
We can probably change it. My guess without looking deeply into it, is that it will be very easy to call them in a lexicographic order (by name of the provider), and it will be somewhat harder to call them in the order in which they are defined in the config. Just based on the way we parse the config, we don't preserve the order. 
If there is an agreement on a lexicographic order, I can easily do that.

**Comment by handlespeed** [comment]
Thanks for reply.
So because of current Orleans provider initialization implementation, What is best practice our project has to do? Merge all logic of every provider into one provider to make sure the order? Or other suggestion?

**Comment by Eldar1205** [comment]
If you use a DI container that is integrated into Orleans as an `IServiceProvider`, and that DI container supports decorator registrations (like SimpleInjector which I use) then you can do the following (worked for me):
1. Create new abstraction interface for bootstrap, e.g. `IBootstrapper`
2. Implement the basic bootstrapping behaviour as class implementing the interface, e.g. `BasicBootstrapper`
3. Implement decorators to `IBootstrapper`, each adding another layer of functionality before/after the decoratee
4. Register to Orleans an `IBootstrapProvider` implementation that resolves `IBootstrapper` from the DI container and executes it. That way you control the bootstrapping order, don't have strong dependency over the framework, keep your bootstrapping behaviours loosely coupled from each other and simple (SOLID design). 

Additionally, SimpleInjector supports collections registrations that preserve order which saves the need for decorators, and several other DI containers out there support that as well.

**Comment by handlespeed** [comment]
Thanks for reply.
Seems your solution is also to use only one bootstrapper provider of Orleans.
Our current providers contain not only DI provider, but also other non-DI related providers which may need DI registration. We are thinking how to merge them into one thing: other provider works after DI registration.

**Comment by sergeybykov** [comment]
@gabikliot 
>My guess without looking deeply into it, is that it will be very easy to call them in a lexicographic order (by name of the provider), and it will be somewhat harder to call them in the order in which they are defined in the config.

Can't we put them in a list in the order they are defined instead of a dictionary?

**Comment by gabikliot** [comment]
@sergeybykov , we can, we will just need to make sure we preserve that order from the time config is read and there is also a different logic to find and load the providers. It may all work fine easily and nicely, or maybe not so easily. I was not sure from a fast glance.
But any of the 2 should probably be doable.

**Comment by sergeybykov** [comment]
My hunch is that it should be fairly easy to preserve the order in which providers get added programmatically or list in config.

**Comment by benjaminpetit** [comment]
Fixed in 2.0 with DI and Lifecycle
